autobahn/wamp_session.ipp

```diff
@@ -231,71 +234,56 @@ boost::future<uint64_t> wamp_session<IStream, OStream>::join(
 
-    packer.pack(static_cast<int> (message_type::HELLO));
-    packer.pack(realm);
+    std::unordered_map<std::string, bool> caller_features;
+    caller_features["call_timeout"] = true;
+    std::unordered_map<std::string, msgpack::object> caller;
+    caller["features"] = msgpack::object(caller_features, zone);
+    roles["caller"] = msgpack::object(caller, zone);
 
+    std::unordered_map<std::string, bool> callee_features;
+    callee_features["call_timeout"] = true;
+    std::unordered_map<std::string, msgpack::object> callee;
+    callee["features"] = msgpack::object(callee_features, zone);
+    roles["callee"] = msgpack::object(callee, zone);
 
-    // set authentication data - if any given
-    if ( !authmethods.empty() && !authid.empty() ) {
+    std::unordered_map<std::string, msgpack::object> publisher;
+    roles["publisher"] = msgpack::object(publisher, zone);
 
-    	packer.pack_map(3);
-
-    	// an "authmethods" entry -> [ "wampcra" , "ticket", ... ]
-    	packer.pack(std::string("authmethods"));
-    	packer.pack_array( authmethods.size() );
-        for ( std::string am : authmethods ) {
-    	    packer.pack( am );
-        }
-
-    	// authid -> "principal"
-    	packer.pack(std::string("authid"));
-    	packer.pack(std::string( authid ));
-    }
-    else {
-    	packer.pack_map(1);
-    }
+    std::unordered_map<std::string, msgpack::object> subscriber;
+    roles["subscriber"] = msgpack::object(subscriber, zone);
 
+    std::unordered_map<std::string, msgpack::object> details;
+    details["roles"] = msgpack::object(roles, zone);
+    details["authmethods"] = msgpack::object(authmethods, zone);
+    details["authid"] = msgpack::object(authid, zone);
 
-    // and "roles" entry -> { ..... }
-    packer.pack(std::string("roles"));
-
-    packer.pack_map(4);
-    packer.pack(std::string("caller"));
-    packer.pack_map(1);
-    packer.pack("features");
-    packer.pack_map(1);
-    packer.pack("call_timeout");
-    packer.pack(true);
-    packer.pack(std::string("callee"));
-    packer.pack_map(1);
-    packer.pack("features");
-    packer.pack_map(1);
-    packer.pack("call_timeout");
-    packer.pack(true);
-    packer.pack(std::string("publisher"));
-    packer.pack_map(0);
-    packer.pack(std::string("subscriber"));
-    packer.pack_map(0);
+    // [HELLO, Realm|uri, Details|dict]
+    auto message = std::make_shared<wamp_message>(3, std::move(zone));
+    message->set_field(0, static_cast<int>(message_type::HELLO));
+    message->set_field(1, realm);
+    message->set_field(2, details);
 
     auto weak_self = std::weak_ptr<wamp_session>(this->shared_from_this());
 
-    m_io.dispatch([=]() {
+    m_io_service.dispatch([=]() {
         auto shared_self = weak_self.lock();
         if (!shared_self) {
             return;
         }
 
         if (m_session_id) {
-            throw protocol_error("session already joined");
+            m_session_join.set_exception(protocol_error("session already joined"));
```

Comment:
It seems you can add a line "Fixes #90" to the commit message, and it will auto-close issue #90. Likely also #85.